homeassistant/components/fujitsu_hvac/coordinator.py

```diff
@@ -0,0 +1,57 @@
+"""Coordinator for Fujitsu HVAC integration."""
+from asyncio import gather
+from datetime import timedelta
+import logging
+
+from ayla_iot_unofficial import AylaApi, AylaAuthError
+from ayla_iot_unofficial.fujitsu_hvac import FujitsuHVAC
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import API_REFRESH_SECONDS
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FujitsuHVACCoordinator(DataUpdateCoordinator):
+    """Coordinator for Fujitsu HVAC integration."""
+
+    def __init__(self, hass: HomeAssistant, api: AylaApi) -> None:
+        """Initialize coordindtor for Fujitsu HVAC integration."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Fujitsu HVAC data",
+            update_interval=timedelta(seconds=API_REFRESH_SECONDS),
+        )
+        self.api = api
+
+    async def _async_update_data(self) -> dict[str, FujitsuHVAC]:
+        """Fetch data from api endpoint."""
+        listening_entities = set(self.async_contexts())
+        try:
+            if self.api.token_expired:
+                await self.api.async_sign_in()
+
+            if self.api.token_expiring_soon:
+                await self.api.async_refresh_auth()
+
+            devices = await self.api.async_get_devices()
+        except AylaAuthError as e:
+            raise ConfigEntryAuthFailed("Credentials expired for Ayla IoT API") from e
+
+        if len(listening_entities) == 0:
+            devices = list(filter(lambda x: isinstance(x, FujitsuHVAC), devices))
+        else:
+            devices = list(
+                filter(lambda x: x.device_serial_number in listening_entities, devices)
+            )
```

Comment:
So we only update the device list with devices we already know - unless we have no entities set up to begin with, then we add everything? Why do we do this? Why can't we just add all devices and in the future dynamically add the other entities

Reply:
Not exactly. If there are no devices using the coordinator (which would be the case when we call async_config_entry_first_refresh) then we collect data for all devices so that each device exists when we call async_setup_entry for each platform. Then listening_entities will only contain devices that are interested in getting the updates from the coordinator. Each device.async_update makes 2 api calls so it's worth it to skip devices that aren't listening for updates.